beetsplug/fish.py

```diff
@@ -13,329 +14,393 @@
+    # parse past them and find subcommands in the command-line.
+    """
+    # An 'argparse' description for global options for 'beet'.
+    function __fish_beet_global_optspec
+        string join \\n v/verbose h/help c/config= l/library= d/directory= \\
+            format-item= format-album=
     end
-    return 1
-end
-
-function __fish_beet_using_command
-    set cmd (commandline -opc)
-    set needle (count $cmd)
-    if test $needle -gt 1
-        if begin test $argv[1] = $cmd[2];
-            and not contains -- $cmd[$needle] $FIELDS; end
-                return 0
+    """,
+    # Parse the global command-line options supported by Beets to determine
+    # whether a (specific or arbitrary) subcommand has been specified.  If the
+    # same command-line (up to the subcommand) is seen twice, a cached result
+    # is used.
+    #
+    # If a single argument is provided, the current command-line buffer is
+    # checked to contain a beets subcommand of the given name.  Otherwise, the
+    # current command-line buffer is checked for any beets subcommand.  If the
+    # respective check succeeds, 0 is returned.
+    """
+    # Test for a (particular) 'beet' subcommand in the command-line buffer.
+    function __fish_beet_subcommand
+        set -l cmd (commandline -opc)
+        set -e cmd[1]
+        set -l test_cmd $argv[1]
+
+        set -f cached $__fish_beet_subcommand_cache
+        set -l crange "1 .. $(count $cached)"
+        if not set -q __fish_beet_subcommand_cache
+            or test "$cmd[$crange]" != "$cached"
+
+            argparse --stop-nonopt (__fish_beet_global_optspec) -- $cmd
+            or return 1
+
+            set -q _flag_help; and return 1
+            test (count $argv) -eq 0; and return 1
+
+            set -l crange "1 .. -$(count $argv)"
+            set -g __fish_beet_subcommand_cache $cmd[$crange]
+            set -f cached $__fish_beet_subcommand_cache
+        end 2>/dev/null
+
+        if test -n "$test_cmd"
+            test "$test_cmd" = "$cached[-1]"
+        else
+            test (count $cached) -gt 0
         end
     end
-    return 1
-end
-
-function __fish_beet_use_extra
-    set cmd (commandline -opc)
-    set needle (count $cmd)
-    if test $argv[2]  = $cmd[$needle]
-        return 0
+    """,
+    # Determine whether a metadata field-value pair is being filled out, for a
+    # particular field name.  This is used for completions in search queries.
+    """
+    # Test for a '<field>:<value>' argument in the command-line buffer.
+    function __fish_beet_metadata_param
+        set -l cmd (commandline -ct)
+        set -l field (string split -f 1 -- ":" $cmd)
+        or return 1
+
+        if test -n "$argv[1]"
+            test "$field" = "$argv[1]"
+        else
+            return 0
+        end
     end
-    return 1
-end
-"""
+    """,
+]
+
+
+def fish_config_dir() -> Path:
+    """
+    The directory for the user's Fish configuration.
+    """
+
+    config_home: Path
```

Comment:
Consider using `platformdirs`

Reply:
Oh, shoot.  I completely forgot about the `platformdirs` PR I was supposed to open.  Do you mind if I leave this as-is for now and address it in that PR?

Reply:
Sounds good. You may be able to drop this altogether if you decide to print the completions to stdout instead of saving them to a file 😉